### PR TITLE
Ft like dislike article 167313408

### PR DESCRIPTION
**Pull request by salviosage**
#### What does this PR do?

- clap and dislike and unlike functionality

#### Description of Task to be completed?

- user should be able to clap many times for a certain article which he doesn't own

- user should be able to revert that claps to 0 by unlike

- user should be able to dislike article which he doesn't own 

#### How should this be manually tested?
- by cloning this repo swich to this branch and 

- send a GET request to /likes/claps/`article slug` to get all claps to that article

- send a GET request to /likes/dislikes/`article slug` to get all dislikes to that article

- send a PUT request to /likes/clap/`article slug` to create clap for that article 

- send a PUT request to /likes/clap/`article slug` to create clap again to that article 

- send a PUT request to /likes/dislike/`article slug` to dislike that article
- send a PUT request again to /likes/dislike/`article slug` to  unlike article

- send a PUT request to /likes/unlike/`article slug` to unlike that article

#### What are the relevant pivotal tracker stories?
[#167313408](https://www.pivotaltracker.com/story/show/167313408)

#### Screenshots (if appropriate)
Screen Shot 2019-08-19 at 23 41 52
Screen Shot 2019-08-19 at 23 47 47
Screen Shot 2019-08-19 at 23 48 44
Screen Shot 2019-08-19 at 23 49 19
Screen Shot 2019-08-19 at 23 49 45
Screen Shot 2019-08-19 at 23 51 54
Screen Shot 2019-08-19 at 23 52 58
Screen Shot 2019-08-19 at 23 53 49
Screen Shot 2019-08-19 at 23 54 34

### CODE REVIEW

File: src/routes/api/likes/likes.routes.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+import validateToken from '../../../middlewares/auth';
+import confirmEmaiAuth from '../../../middlewares/emailVarification.middleware';
+import isLike from '../../../middlewares/likes.middleware';
+
+const router = express.Router();
+// greate edit a viewing profile handlers
+
+router.get('/claps/:Article', [validateToken, confirmEmaiAuth, isLike], LikesController.getClaps);
+router.get('/dislikes/:Article', [validateToken, confirmEmaiAuth, isLike], LikesController.getDislikes);
+router.put('/clap/:Article', [validateToken, confirmEmaiAuth, isLike], LikesController.clap);
```
Comment (minega25) [inline src/routes/api/likes/likes.routes.js:12]: @salviosage I suggest we maintain the same naming convention for our routes ie `clap` to `claps`

File: src/routes/api/likes/likes.routes.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+router.put('/dislike/:Article', [validateToken, confirmEmaiAuth, isLike], LikesController.dislike);
```
Comment (minega25) [inline src/routes/api/likes/likes.routes.js:13]: Same thing here as well

File: src/routes/index.js [hunk 3]
```diff
@@ -38,9 +39,10 @@ const apiVersion = process.env.API_VERSION;
 const baseUrl = `/api/${apiVersion}`;
 
 router.get('/', (req, res) => res.status(200).json({ status: 200, data: 'Welcome to Authors Haven.' }));
+router.use(article);
 router.use(baseUrl, api);
 router.use(oauth);
-router.use(article);
+router.use(likes);
```
Comment (minega25) [inline src/routes/index.js]: Why are we having likes route outside the baseurl format we are using for our API?
We are now having 10 routes instead of 5 you intended. 
ie those with (localhost:3000/api/v1/likes/.......) as per `line 20 in src/routes/api/index.route.js`
and those with (localhost:3000/likes/.......) as per `line 45 in src/routes/index.js`

**Review by minega25 — CHANGES_REQUESTED** [review]
Great job @salviosage! Find included comments to make our API neater.🙂

File: test/like.test.js [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+        expect(res).to.be.an('object');
+        expect(res.body).to.have.keys('message', 'status', 'data');
+        expect(res.body.message).to.deep.equal('Successfully claped');
+        expect(res.body.data).to.have.keys('ArticleSlug', 'status', 'claps');
+        expect(res.body.data.claps).to.deep.equal(1);
+        expect(res.body.data.status).to.deep.equal('like');
+        done();
+      });
+  });
+  it('Should successfully clap ', (done) => {
```
Comment (nkalyesubula) [inline test/like.test.js]: This test is a duplicate of the one above. Consider removing it.

File: test/like.test.js [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+        if (error) done(error);
+        expect(res).have.status(200);
+        expect(res).to.be.an('object');
+        expect(res.body).to.have.keys('message', 'data', 'status');
+        expect(res.body.message).to.deep.equal('Successfully dislike retrieved');
+        expect(res.body.data).to.have.keys('dislikes');
+        done();
+      });
+  });
+  it('should Unlike Article due to dislike secomd time ', (done) => {
```
Comment (nkalyesubula) [inline test/like.test.js]: Consider giving a clear description to tests

File: test/like.test.js [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+        expect(res).to.be.an('object');
+        expect(res.body).to.have.keys('message', 'status', 'data');
+        expect(res.body.message).to.deep.equal('Successfully disliked');
+        expect(res.body.data).to.have.keys('ArticleSlug', 'status', 'claps');
+        expect(res.body.data.claps).to.deep.equal(0);
+        expect(res.body.data.status).to.deep.equal('dislike');
+        done();
+      });
+  });
+  it('should dislike Article', (done) => {
```
Comment (nkalyesubula) [inline test/like.test.js]: This test is a duplicate of the one above

File: test/like.test.js [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+        expect(res.body).to.have.keys('message', 'status', 'data');
+        expect(res.body.message).to.deep.equal('Successfully unliked');
+        expect(res.body.data).to.have.keys('ArticleSlug', 'status', 'claps');
+        expect(res.body.data.claps).to.deep.equal(0);
+        expect(res.body.data.status).to.deep.equal('neutral');
+        done();
+      });
+  });
+
+  it('should dislike Article', (done) => {
```
Comment (nkalyesubula) [inline test/like.test.js:164]: The test before this says 'should unlike Article' and this one says 'should dislike Article'. What is the difference between these two tests, and why do we have both of them?

File: test/like.test.js [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+      })
+      .end((error, res) => {
+        if (error) done(error);
+        mikeToken = `Bearer ${res.body.token}`;
+        expect(res.status).to.be.equal(200);
+        expect(res.body).to.have.deep.property('message');
+        done();
+      });
+  });
+  it('should login', (done) => {
```
Comment (nkalyesubula) [inline test/like.test.js]: What's the difference between 'should login and 'should login mike'?? And these are two separate tests

File: src/routes/index.js (hunk 3 above)
Comment (nkalyesubula) [inline src/routes/index.js]: Consider adding a baseUrl to this route

**Review by nkalyesubula — CHANGES_REQUESTED** [review]
Thank you, Salvi, for the hard work. I have realized though that you have many duplicates in your tests, and your routes are not including the baseUrl, kindly work on this

File: test/like.test.js (hunk 4 above)
Comment (salviosage) [reply test/like.test.js]: > This test is a duplicate of the one above. Consider removing it.

thanks, noa for your feedback however I have to test if someone can clap so many times that  means I have to clap multiple time

File: src/app.js [hunk 9]
```diff
@@ -7,7 +7,7 @@ import { mock } from './middlewares/validators/socialLogin-mock';
 
 const app = express();
 
-if (process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'test') {
+if (process.env.NODE_ENV === 'test' || process.env.NODE_ENV === 'development') {
```
Comment (Mnickii) [inline src/app.js:10]: Is this change necessary?

File: src/controllers/likes.controller.js [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+   *  @static
+   */
+  static async getDislikes(req, res) {
+    try {
+      const dislike = await likeService.getAllADislike(req.params.Article);
+      if (dislike) {
+        const data = {
+          dislikes: dislike.count,
+        };
+        util.setSuccess(200, 'Successfully dislike retrieved', data);
```
Comment (Mnickii) [inline src/controllers/likes.controller.js]: Dislike retrieved successfully would be a better message

File: src/controllers/likes.controller.js [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+   */
+  static async getClaps(req, res) {
+    try {
+      const claps = await likeService.getAllAClaps(req.params.Article);
+      if (claps) {
+        const data = {
+          clapers: claps.count,
+          claps: Object.values(claps)[1],
+        };
+        util.setSuccess(200, 'Successfully claps retrieved', data);
```
Comment (Mnickii) [inline src/controllers/likes.controller.js]: Same here

**Review by Mnickii — CHANGES_REQUESTED** [review]

File: test/like.test.js (hunk 5 above)
Comment (salviosage) [reply test/like.test.js]: > Consider giving a clear description to tests

worked on it

File: test/like.test.js (hunk 7 above)
Comment (salviosage) [reply test/like.test.js:164]: Unliking article is setting like to neutral, dislike is setting like status to dislike .
user can either clap , dislike and unlike an article but you can't unlike article which you didnt like or dislike before. hope thats is clear Noa .

File: test/like.test.js (hunk 8 above)
Comment (salviosage) [reply test/like.test.js]: yes, i need more than one user for me to test all senario.

File: src/app.js (hunk 9 above)
Comment (salviosage) [reply src/app.js:10]: > Is this change necessary?

no, it's not necessary I was testing how the article test is passing which is magic and I realized that test never return any response you can even kill something in test and still running which is kind of magic.  maybe Christian can take his time and explain how things are running

File: src/controllers/likes.controller.js (hunk 10 above)
Comment (salviosage) [reply src/controllers/likes.controller.js]: fixed

File: src/controllers/likes.controller.js (hunk 11 above)
Comment (salviosage) [reply src/controllers/likes.controller.js]: fixed

File: src/app.js (hunk 9 above)
Comment (habinezadalvan) [reply src/app.js:10]: I explained its functionality which works when it comes to mocking the third party (cloudinary); however it seems like the mocking doesn't reach the controller. As Mike's PR is not yet merged we are going to find a way we can access the controller while testing. Before I used mocking tests, tests were reaching the controller, if mocking still brings issues with other tests I think I can reuse normal test. Thank you.

**Review by nkalyesubula — APPROVED** [review]

**Review by minega25 — APPROVED** [review]

**Review by habinezadalvan — CHANGES_REQUESTED** [review]
I resolved the issue of the test and the cloudinary. I think it would be better if you include the changes in your codebase and the rest of us we rebase when you PR is merged.

**Comment by salviosage** [comment]
> I resolved the issue of the test and the cloudinary. I think it would be better if you include the changes in your codebase and the rest of us we rebase when you PR is merged.

nice work @habinezadalvan, since that is not an issue for my Feature I suggest if you can create a bug story to fix the issue.

**Review by Mnickii — APPROVED** [review]
